File: RB110/ruby_interview_problems/spot_question_set/problem3.rb
=begin
Return substring instance count
Complete the solution so that it returns the number of times the search_text 
is found within the full_text.

Usage example:
solution('aa_bb_cc_dd_bb_e', 'bb') # should return 2 since bb shows up twice
solution('aaabbbcccc', 'bbb') # should return 1

Problem: given two string arguments, return an integer

Rules:
- First string arugment is the entire text
- Second string arg is a substring
- Return integer counts the occurrenes of the substring within the full string

Algorithm:
- CREATE an array of all possible substrings
- ITERATE over array of substrings and select the substring if it 
  is equal to the target substring
- RETURN the count of the selected array
=end


def solution(string, target)
  substr = []

  string.size.times do |start_idx|
    start_idx.upto(string.size - 1) do |end_idx|
       substr << string[start_idx..end_idx]
    end
  end

  substr.select { |subs| subs == target }.count

end


p solution('abcdeb','b') == 2
p solution('abcdeb', 'a') == 1
p solution('abbc', 'bb') == 1
p solution('abcdeb', 'bd') == 0

File: LS220/DSA_small_problems/recursion/problem3.js
/*

Create a recursive function that counts the number of digits in a given integer. The function should accept an integer and return the count of its digits. For instance, if the input is 12345, the function should return 5.

--- Problem
INPUT: integer
OUTPUT: integer

  Rules:
  Given an integer
  return an integer where
    - represents the number of digits in the given integer argument
    - find count of digits recurisvely

  - Base case:
    - If given an integer under 10, I return 1
  - Recursive definition:
    - The count of digits in an integer is the first digit plus the count of the rest of the digits in the integer
*/

function countDigits(number) {
  if (number < 10 && number > -10) return 1;
  let count = 1;

  return count + countDigits(Math.floor(number / 10));
}


// Example test cases:
console.log(countDigits(-100000) === 6);
console.log(countDigits(-12345) === 5);
console.log(countDigits(-7) === 1);
console.log(countDigits(12345) === 5);
console.log(countDigits(7) === 1);
console.log(countDigits(100000) === 6);
console.log(countDigits(99999) === 5);
console.log(countDigits(0) === 1);

// All test cases should log true.